File: Authentication/src/main/java/authentication/ports/AuthenticationService.java
package authentication.ports;

import java.util.Optional;
import authentication.domain.model.*;
import authentication.domain.exceptions.*;

public interface AuthenticationService {

	public String register(Username username, String password)
			throws UserAlreadyExistsException, SomethingWentWrongException;

	public String authenticate(Username username, String password)
			throws UserNotFoundException, WrongCredentialsException;

	public String refresh(String jwt)
			throws PasswordAuthenticationRequiredException, SessionExpiredException, InvalidTokenException;

	public Username validate(String jwt) throws SessionExpiredException, InvalidTokenException;

	public void forceAuthentication(String jwt, Username username)
			throws UserNotFoundException, SessionExpiredException, InvalidTokenException;

	public Optional<String> healthCheckError();
}

File: ApiGateway/src/main/scala/apigateway/ports/AuthenticationService.scala
package apigateway.ports

import scala.concurrent.Future
import apigateway.domain.model.*

trait AuthenticationService:
  import AuthenticationService.*
  def register(
      dto: RegisterUserDTO
  ): Future[Either[UserAlreadyExists | SomethingWentWrong, String]]

  def authenticate(
      username: Username,
      dto: AuthenticateUserDTO
  ): Future[Either[UserNotFound | UnauthorizedError, String]]

  def refresh(
      bearertoken: String
  ): Future[Either[BadAuthorizationHeader | UnauthorizedError, String]]

  def validate(
      bearertoken: String
  ): Future[Either[BadAuthorizationHeader | UnauthorizedError, Boolean]]

  def forceAuthentication(
      bearertoken: String,
      username: Username
  ): Future[
    Either[BadAuthorizationHeader | UserNotFound | UnauthorizedError, Unit]
  ]

object AuthenticationService:
  case class UserAlreadyExists(username: Username)
  case class UserNotFound(username: Username)
  case class SomethingWentWrong(message: String)
  case class BadAuthorizationHeader()
  case class UnauthorizedError(message: String)
